### PR TITLE
Add style encoding for strings

### LINKED ISSUE #3992: Use StringEncoding in TextManager
This issue introduces the `StringEncoding` described in the following design doc: https://hackmd.io/Ybil_epnQPiBczPcbGRqCw?view

It adds the ability to define constant, manual, direct, and formatted string encodings in `TextManager`. This is related to #3053 and should close it when done.

### PR DESCRIPTION
**Pull request by andy-sweet**
# Description
This is the beginning of the visualization style implementation described in the following doc: https://hackmd.io/Ybil_epnQPiBczPcbGRqCw

It introduces the generic `StyleEncoding` protocol. The public API of this protocol is `__call__`, which is stateless, and encodes style values from layer feature. This generic protocol should support strings, colors, and any other value that can be stored in array-likes.

This protocol also has a private API that is stateful and needs to be implemented to support napari's current behavior of storing and exposing style values (e.g. `Points.text.values`). We could consider splitting the private API into a separate `StyleStore` class or something similar, but I've struggled to make that work well with the rest of napari in experimental PRs. As it's private and if we discourage people from implementing the protocol, then I think it's OK to handle that later if needed/desired.

My long term vision is that only the `__call__` public API will remain, and the values will be stored on immutable views/slices of a layer. But that is still a little way off.

It also introduces the specific `StringEncoding` that encodes strings from features. The public API of that specific type of encoding is covered by tests. 

For more context on how this will be used to implement `TextManager` without any significant changes in behavior, see the following draft PR in my fork: https://github.com/andy-sweet/napari/pull/2

Both new modules are private to start (i.e. with a leading underscore) to avoid expanding the napari public API unnecessarily, but at least `string_encoding` will be made public when it is used by `TextManager`.

A few issues/questions to highlight here.
- I used `DirectStringEncoding` to represent a string that comes directly from a specific features column and `ManualStringEncoding` to represent a string that is manually specified in an array. That's in conflict with the use of `DIRECT` mode in some places now, which means manual, but I think it's a worthwhile improvement.
- The generic types are bound to be subtypes of `np.ndarray`. This makes sense for colors, but maybe less sense for strings. It makes things a little simpler, but I also think it wouldn't be difficult to define `StringValue = str`.
- `__call__` returns either a single style value or multiple style values in an array. Previously I always returned an array that matched the length of the given features, but there is at least one case where that's inefficient in general (i.e. constant encodings), so I'd like to also allow returning the single value as it can easily broadcast to the array.
- I did not write tests for the private API yet. We could do that here, but I think that testing through `test_text_manager`, `test_points`, etc. in follow-up PRs should be sufficient.
- I'm not sure where these new modules should live. `napari.layers.utils` seems reasonable, but maybe they should go in `napari.utils`?

## Type of change
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
Part of #3992.

# How has this been tested?
- [x] example: all tests in `test_string_encoding.py` pass

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by andy-sweet** [comment]
@jni : I hope you'll be interested in this. Also, no manager classes this time ;)

File: napari/layers/utils/_string_encoding.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+    Parameters
+    ----------
+    value : StringEncodingArgument
+        The value to validate and coerce.
+        If this is already a StringEncoding, it is returned as is.
+        If this is a dict, then it should represent one of the built-in string encodings.
+        If this a valid format string, then a FormatStringEncoding is returned.
+        If this is any other string, a ConstantStringEncoding is returned.
+        If this is a sequence of strings, a ManualStringEncoding is returned.
```
Comment (jni) [inline napari/layers/utils/_string_encoding.py:141]: ❤️

File: napari/layers/utils/_style_encoding.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+StyleArray = TypeVar('StyleArray', bound=np.ndarray)
+
+
+@runtime_checkable
+class StyleEncoding(Protocol[StyleValue, StyleArray]):
+    """Encodes style values, like colors and strings, from layer features.
+
+    The public API of any StyleEncoding is just __call__, such that it can
+    be called to generate style values from layer features. That call should
+    be stateless, in that the values returned only depend on the given features.
```
Comment (jni) [inline napari/layers/utils/_style_encoding.py:26]: Also ❤️

**Review by jni — APPROVED** [review]
Sorry for all the waiting @andy-sweet! I literally have zero suggestions though! 😃 As you already noted, to me it's a bit weird that StringValue is an array, but let's see how we go with it, if you think it makes things simpler. Overall I feel that this represents very well the ideal that we've been chasing of a strict protocol hierarchy together with convenience functions to allow flexible inputs from users. I love it! ❤️ 🚀

**Comment by sofroniewn** [comment]
Thanks for the review @jni - @andy-sweet onwards and upwards! 🚀

**Comment by tlambert03** [comment]
Just wanted to chime in and also say thanks for this PR @andy-sweet.  I did look at it a couple times and didn't feel up enough on the topic to say much :) so I never actually posted a review.  The typing is lovely, your explanatory documents are awesome.  I'm perfectly happy moving forward with your lead on this!

**Comment by andy-sweet** [comment]
Thanks for the reviews!

**Pull request merged**